src/tools/wasm-ctor-eval.cpp

```diff
@@ -489,27 +489,38 @@ struct CtorEvalExternalInterface : EvallingModuleInstance::ExternalInterface {
   }
 };
 
+struct EvalCtorOutcome {
+  // Whether we completely evalled the function (that is, we did not fail, and
+  // we did not only partially eval it).
+  bool evalledCompletely;
+
+  // If the function was evalled completely, and it returns something, that
+  // value is given here.
+  Literals results;
```

Comment:
It could, yes. But there are 3 states here:

1. We failed to eval everything completely. There are no results even if the function returns something.
2. We succeeded. The function has no results.
3. We succeeded. The function has results, and the results are provided.

We could use `std::optional<Literals>` and use a `Literal` of type `none` to mean "succeeded, no results". (Or perhaps an empty Literals, although I'm not sure if that's valid?)

Overall I think given the 3 states it's nicer to be explicit which is why I added a struct. But if it feels like overkill I can remove it.

Reply:
Let me land this as is for now, as it blocks further work. But happy to do a followup to change this if you want.

Reply:
I was imagining that case 2 would be covered by an empty `Literals`, which is what the interpreter returns when executing functions with no return value.

Reply:
Yeah, that could work, I agree. It feels not explicit enough to me, but I don't feel strongly. What do you prefer?

Reply:
Another more explicit option would be to have a `std::variant<EvalFailure, Literals>` rather than a `std::optional<Literals>`, but I think it would be good to use one of those two solutions.

Reply:
Ok, I'll open a PR.